### PR TITLE
Added a new tool to merge a collection of variant calling metrics

### PR DESCRIPTION
**Pull request by eitanbanks**
### Description

Useful tool to take a bunch of sharded variant metrics files and accumulate them into a single merged version.

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [x] Added or modified tests to cover changes and any new functionality
- [x] Edited the README / documentation (if applicable)
- [x] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

### CODE REVIEW

**Comment by eitanbanks** [comment]
@yfarjoun I assume you want to review this, right?

**Comment by eitanbanks** [comment]
@kcibul here's that PR you asked for

File: src/main/java/picard/vcf/CollectVariantCallingMetrics.java [hunk 1]
```diff
@@ -293,8 +312,16 @@ public void calculateDerivedFields() {
             super.calculateDerivedFields();
             // Divide by zero should be OK -- NaN should get propagated to metrics file.
```
Comment (yfarjoun) [inline src/main/java/picard/vcf/CollectVariantCallingMetrics.java]: ouch. NaNs will bite. need to have a solution for NaNs. can you add a test to see what will happen and protect against it?

Comment (yfarjoun) [inline src/main/java/picard/vcf/CollectVariantCallingMetrics.java]: In case of Nan, denominator must have been zero....so that's easy.

File: src/main/java/picard/vcf/AccumulateVariantCallingMetrics.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                final CollectVariantCallingMetrics.VariantCallingSummaryMetrics summaryMetrics = summaryMetricsFile.getMetrics().get(0);
+                summaryMetrics.calculateFromDerivedFields(totalHetDepth);
+                summaries.add(summaryMetrics);
+            } catch (IOException e) {
+                throw new PicardException(String.format("Cannot read from metrics files with prefix %s", inputPrefix));
+            }
+        }
+
+        // now merge all of the accumulated metrics
+        final Collection<CollectVariantCallingMetrics.VariantCallingDetailMetrics> collapsedDetails = new ArrayList<>();
```
Comment (yfarjoun) [inline src/main/java/picard/vcf/AccumulateVariantCallingMetrics.java]: Why not add directly to the metric file? seems like an un-needed extra Collection.

File: src/main/java/picard/vcf/AccumulateVariantCallingMetrics.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+import picard.cmdline.StandardOptionDefinitions;
+import picard.cmdline.programgroups.VcfOrBcf;
+
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.*;
+
+/**
+ * Combines multiple Variant Calling Metrics files into a single file.
```
Comment (yfarjoun) [inline src/main/java/picard/vcf/AccumulateVariantCallingMetrics.java]: author here as well?

File: src/main/java/picard/vcf/AccumulateVariantCallingMetrics.java (hunk 2 above)
Comment (eitanbanks) [inline src/main/java/picard/vcf/AccumulateVariantCallingMetrics.java]: I rewrote the code to address this and didn't like the new version.  You either need to do extra compute (to recalculate derived fields) so many unneeded times or you leave the summary collection in a dangerously inconsistent state.  I'd like to advocate that we keep the code as is since it's safer.

File: src/main/java/picard/vcf/CollectVariantCallingMetrics.java [hunk 4]
```diff
@@ -136,23 +136,23 @@ protected int doWork() {
 
     /** A collection of metrics relating to snps and indels within a variant-calling file (VCF). */
     public static class VariantCallingSummaryMetrics extends MergeableMetricBase {
-        /** The number of high confidence SNPs calls (i.e. non-reference genotypes) that were examined */
+        /** The number of passing bi-allelic SNPs calls (i.e. non-reference genotypes) that were examined */
         @MergeByAdding
         public long TOTAL_SNPS;
 
-        /** The number of high confidence SNPs found in dbSNP */
+        /** The number of passing SNPs found in dbSNP */
```
Comment (eitanbanks) [inline src/main/java/picard/vcf/CollectVariantCallingMetrics.java]: Also changed for dbsnp rate and ti/tv ratios below.

File: src/main/java/picard/vcf/CollectVariantCallingMetrics.java (hunk 1 above)
Comment (eitanbanks) [inline src/main/java/picard/vcf/CollectVariantCallingMetrics.java]: I didn't touch this code at all -- that's from a previous commit.  If you want it changed then let's have someone do it in a different PR.

**Comment by eitanbanks** [comment]
@yfarjoun I addressed all of your comments either in the code or the comments above.  Please re-review!

File: src/main/java/picard/vcf/AccumulateVariantCallingMetrics.java [hunk 5]
```diff
@@ -84,9 +87,7 @@ protected int doWork() {
                     totalHetDepth += detailedMetrics.TOTAL_HET_DEPTH;
 
                     // add it to the list of metrics for that sample so that we can merge them later
-                    if (!sampleDetailsMap.containsKey(detailedMetrics.SAMPLE_ALIAS)) {
-                        sampleDetailsMap.put(detailedMetrics.SAMPLE_ALIAS, new ArrayList<>());
-                    }
+                    sampleDetailsMap.computeIfAbsent(detailedMetrics.SAMPLE_ALIAS, f -> new ArrayList<>());
```
Comment (yfarjoun) [inline src/main/java/picard/vcf/AccumulateVariantCallingMetrics.java]: you can merge these two lines. computeIfAbsent will return the eventual object...so:

```
sampleDetailsMap.computeIfAbsent(detailedMetrics.SAMPLE_ALIAS, f -> new ArrayList<>())
.add(detailedMetrics);
```

File: src/main/java/picard/vcf/CollectVariantCallingMetrics.java (hunk 1 above)
Comment (yfarjoun) [inline src/main/java/picard/vcf/CollectVariantCallingMetrics.java]: That's not what I meant. You might have NaNs in the derived metrics from which you try to infer the private metrics. You need to be able to work-around that.

File: src/main/java/picard/vcf/AccumulateVariantCallingMetrics.java (hunk 2 above)
Comment (yfarjoun) [inline src/main/java/picard/vcf/AccumulateVariantCallingMetrics.java]: the metrics file is a collection already...so why not add to that container instead of the one you define here. I don't understand your answer.

Comment (eitanbanks) [inline src/main/java/picard/vcf/AccumulateVariantCallingMetrics.java]: Resolved in person.

**Comment by eitanbanks** [comment]
@yfarjoun I've made the changes we discussed and added another test that includes metrics from a completely empty shard with zeroes and NaNs.
Are we good to go?

**Comment by eitanbanks** [comment]
Okay, @yfarjoun, I've addressed the latest round of nitpicking.  I took the liberty of squashing into a single commit because I'm confident that we're ready to put this baby to bed.  :)

**Comment by yfarjoun** [comment]
So old-fashioned! :-) nowadays github enables you to rebase/squash instead of merging. This is preferred as it enables the reviewer to easily see what were your most recent changes...

**Comment by eitanbanks** [comment]
Back when I was an engineer github had no such capabilities...

File: src/main/java/picard/vcf/AccumulateVariantCallingMetrics.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+/**
+ * Combines multiple Variant Calling Metrics files into a single file.
+ * @author Eric Banks
+ */
+@CommandLineProgramProperties(
+        usage = "Combines multiple Variant Calling Metrics files into a single file.  This tool is used in cases where the metrics are calculated" +
+                " separately for different (genomic) shards of the same callset and we want to combine them into a single result over the entire callset." +
+                " The shards are expected to contain the same samples (although it will not fail if they do not) and to not have been run over overlapping genomic positions.",
+        usageShort = "Combines multiple Variant Calling Metrics files into a single file",
+        programGroup = VcfOrBcf.class
```
Comment (yfarjoun) [inline src/main/java/picard/vcf/AccumulateVariantCallingMetrics.java]: sorry I didn't see this before. 

This should be in the `Metrics` class, not `VcfOrBcf`

**Comment by eitanbanks** [comment]
okay, made and pushed those changes.  Will wait for tests to pass and then merge.

**Comment by yfarjoun** [comment]
🎉

**Comment by kbergin** [comment]
To clean up this repo, we are deleting stale branches that have been merged or closed. You have a short period of time to restore this branch if needed. Thanks!
